File: packages/server/cmd/renewlet/ai_recognition_notes.go
package main

import (
	"net/url"
	"strings"
	"unicode"
)

var aiRecognitionProcessNoteFragments = []string{
	"ai根据",
	"ai建议",
	"ai未能",
	"ai无法",
	"ai生成",
	"ai猜测",
	"输入没有",
	"输入未",
	"用户输入",
	"原文",
	"图片未",
	"图像未",
	"表格未",
	"无法确定",
	"不能确定",
	"不确定",
	"未能确定",
	"未能高置信识别",
	"无法识别",
	"未明确",
	"未提供",
	"没有提供",
	"请确认",
	"需要确认",
	"确认后",
	"低置信",
	"高置信识别",
	"模型返回",
	"模型输出",
	"模型无法",
	"aigenerated",
	"aiguessed",
	"theinputdoesnot",
	"inputdoesnot",
	"inputdidnot",
	"notprovided",
	"notspecified",
	"uncertain",
	"cannotdetermine",
	"cantdetermine",
	"pleaseconfirm",
	"lowconfidence",
	"modeloutput",
	"modelreturned",
}

var aiRecognitionNoteDropClauseFragments = []string{
	"适合记录",
	"可用于记录",
	"用于记录",
	"便于记录",
	"方便记录",
	"订阅管理",
	"套餐订阅",
	"请确认",
	"需要确认",
	"导入",
	"renewlet",
	"suitableforrecording",
	"canbeusedtorecord",
	"usedtorecord",
	"subscriptionmanagement",
	"pleaseconfirm",
	"import",
}

var aiRecognitionNoteMarketingFragments = []string{
	"优质",
	"领先",
	"专业",
	"全方位",
	"一站式",
	"可靠",
	"高性能",
	"稳定",
	"premium",
	"leading",
	"professional",
	"comprehensive",
	"all-in-one",
}

func missingDescribableAINoteNames(subscriptions []aiRecognizedSubscriptionDraft) []string {
	names := []string{}
	for _, draft := range subscriptions {
		if draft.Notes == nil && isDescribableForAINotes(draft) {
			names = append(names, draft.Name)
			if len(names) >= 20 {
				break
			}
		}
	}
	return names
}

func isDescribableForAINotes(draft aiRecognizedSubscriptionDraft) bool {
	return draft.Website != nil || draft.Category != nil || len(draft.Tags) > 0 || draft.Confidence == "high"
}

func fillMissingAINotesWithDynamicFallback(response aiRecognizeResponse, locale appLocale, configContext aiRecognitionConfigContext) aiRecognizeResponse {
	for index := range response.Subscriptions {
		draft := &response.Subscriptions[index]
		if draft.Notes != nil || !isDescribableForAINotes(*draft) {
			continue
		}
		note := buildDynamicAIFallbackNote(*draft, locale, configContext)
		if note == "" {
			continue
		}
		draft.Notes = &aiSuggestedTextField{Value: note, Source: "suggested"}
		filtered := []string{}
		for _, warning := range draft.Warnings {
			if warning != "AI_WARNING_NOTES_MISSING" {
				filtered = append(filtered, warning)
			}
		}
		draft.Warnings = compactAIWarnings(filtered, 12)
	}
	return response
}

func buildDynamicAIFallbackNote(draft aiRecognizedSubscriptionDraft, locale appLocale, configContext aiRecognitionConfigContext) string {
	labels := dynamicAINoteLabels(draft, configContext)
	var raw string
	if len(labels) > 0 {
		raw = serverFormat(locale, "aiRecognition.fallbackNote.labels", map[string]interface{}{
			"name":   draft.Name,
			"labels": strings.Join(labels, serverText(locale, "aiRecognition.fallbackNote.labelSeparator")),
		})
	} else {
		raw = dynamicAIWebsiteFallbackNote(draft, locale)
	}
	field := normalizeAINotesField(&aiSuggestedTextField{Value: raw, Source: "suggested"}, 5000)
	if field == nil {
		return ""
	}
	return field.Value
}

func dynamicAINoteLabels(draft aiRecognizedSubscriptionDraft, configContext aiRecognitionConfigContext) []string {
	labels := []string{}
	seen := map[string]bool{}
	add := func(value string) {
		label := trimMax(value, 80)
		key := aiRecognitionConfigMatchKey(label)
		if label == "" || key == "" || key == aiRecognitionConfigMatchKey(draft.Name) || seen[key] {
			return
		}
		seen[key] = true
		labels = append(labels, label)
	}
	for _, tag := range draft.Tags {
		add(tag)
	}
	if draft.Category != nil {
		if option := findAIRecognitionConfigOption(configContext.Categories, *draft.Category); option != nil {
			add(option.Label)
		} else {
			add(*draft.Category)
		}
	}
	if len(labels) > 3 {
		return labels[:3]
	}
	return labels
}

func dynamicAIWebsiteFallbackNote(draft aiRecognizedSubscriptionDraft, locale appLocale) string {
	if draft.Website == nil {
		return ""
	}
	host := aiRecognitionHostname(draft.Website.Value)
	if host == "" {
		return ""
	}
	return serverFormat(locale, "aiRecognition.fallbackNote.website", map[string]interface{}{
		"name":   draft.Name,
		"domain": host,
	})
}

func aiRecognitionHostname(value string) string {
	text := strings.TrimSpace(value)
	if text == "" {
		return ""
	}
	if !strings.HasPrefix(text, "http://") && !strings.HasPrefix(text, "https://") {
		text = "https://" + text
	}
	parsed, err := url.Parse(text)
	if err != nil || parsed.Hostname() == "" {
		return ""
	}
	return strings.TrimPrefix(parsed.Hostname(), "www.")
}

func normalizeAINotesField(field *aiSuggestedTextField, maxLen int) *aiSuggestedTextField {
	normalized := normalizeAISuggestedField(field, maxLen)
	if normalized == nil {
		return nil
	}
	// AI 备注会进入订阅长期备注；这里兜底清掉识别过程和产品内视角，保留真实服务/网站简介。
	if isAIRecognitionProcessNote(normalized.Value) {
		return nil
	}
	value := stripAIRecognitionAdviceClauses(normalized.Value)
	if value == "" || isAIRecognitionProcessNote(value) || isAIRecognitionMarketingNote(value) {
		return nil
	}
	normalized.Value = trimMax(cleanAIRecognitionServiceDescription(value), maxLen)
	if normalized.Value == "" {
		return nil
	}
	return normalized
}

func isAIRecognitionProcessNote(value string) bool {
	key := aiRecognitionNoteMatchKey(value)
	if key == "" {
		return false
	}
	for _, fragment := range aiRecognitionProcessNoteFragments {
		if strings.Contains(key, fragment) {
			return true
		}
	}
	return false
}

func aiRecognitionNoteMatchKey(value string) string {
	var builder strings.Builder
	for _, r := range strings.ToLower(strings.TrimSpace(value)) {
		if unicode.IsSpace(r) {
			continue
		}
		builder.WriteRune(r)
	}
	return builder.String()
}

func stripAIRecognitionAdviceClauses(value string) string {
	clauses := splitAIRecognitionNoteClauses(value)
	kept := []string{}
	dropped := false
	for _, clause := range clauses {
		text := strings.TrimSpace(clause)
		if text == "" || aiRecognitionNoteContainsFragment(text, aiRecognitionNoteDropClauseFragments) {
			if text != "" {
				dropped = true
			}
			continue
		}
		kept = append(kept, text)
	}
	if !dropped {
		return strings.TrimSpace(value)
	}
	return strings.TrimSpace(strings.Join(kept, "，"))
}

func splitAIRecognitionNoteClauses(value string) []string {
	clauses := []string{}
	var builder strings.Builder
	for _, r := range value {
		if strings.ContainsRune("，,。.!?！？；;", r) {
			clauses = append(clauses, builder.String())
			builder.Reset()
			continue
		}
		builder.WriteRune(r)
	}
	clauses = append(clauses, builder.String())
	return clauses
}

func isAIRecognitionMarketingNote(value string) bool {
	return aiRecognitionNoteContainsFragment(value, aiRecognitionNoteMarketingFragments)
}

func aiRecognitionNoteContainsFragment(value string, fragments []string) bool {
	key := aiRecognitionNoteMatchKey(value)
	if key == "" {
		return false
	}
	for _, fragment := range fragments {
		if strings.Contains(key, fragment) {
			return true
		}
	}
	return false
}

func cleanAIRecognitionServiceDescription(value string) string {
	text := strings.ReplaceAll(value, "相关服务", "服务")
	text = strings.ReplaceAll(text, "等服务服务", "等服务")
	text = strings.ReplaceAll(text, "服务服务", "服务")
	return strings.Join(strings.Fields(strings.TrimSpace(text)), " ")
}

File: packages/shared/src/ai-recognition-notes.ts
const AI_RECOGNITION_PROCESS_NOTE_FRAGMENTS = [
  "ai根据",
  "ai建议",
  "ai未能",
  "ai无法",
  "ai生成",
  "ai猜测",
  "输入没有",
  "输入未",
  "用户输入",
  "原文",
  "图片未",
  "图像未",
  "表格未",
  "无法确定",
  "不能确定",
  "不确定",
  "未能确定",
  "未能高置信识别",
  "无法识别",
  "未明确",
  "未提供",
  "没有提供",
  "请确认",
  "需要确认",
  "确认后",
  "低置信",
  "高置信识别",
  "模型返回",
  "模型输出",
  "模型无法",
  "aigenerated",
  "aiguessed",
  "theinputdoesnot",
  "inputdoesnot",
  "inputdidnot",
  "notprovided",
  "notspecified",
  "uncertain",
  "cannotdetermine",
  "cantdetermine",
  "pleaseconfirm",
  "lowconfidence",
  "modeloutput",
  "modelreturned",
] as const;

const AI_RECOGNITION_NOTE_DROP_CLAUSE_FRAGMENTS = [
  "适合记录",
  "可用于记录",
  "用于记录",
  "便于记录",
  "方便记录",
  "订阅管理",
  "套餐订阅",
  "请确认",
  "需要确认",
  "导入",
  "renewlet",
  "suitableforrecording",
  "canbeusedtorecord",
  "usedtorecord",
  "subscriptionmanagement",
  "pleaseconfirm",
  "import",
] as const;

const AI_RECOGNITION_NOTE_MARKETING_FRAGMENTS = [
  "优质",
  "领先",
  "专业",
  "全方位",
  "一站式",
  "可靠",
  "高性能",
  "稳定",
  "premium",
  "leading",
  "professional",
  "comprehensive",
  "all-in-one",
] as const;

/**
 * 清洗 AI 识别草稿中的长期备注。
 *
 * AI 备注会进入订阅长期记录；这里只兜底清掉识别过程、产品内视角和营销套话，保留真实服务/网站简介。
 */
export function normalizeAIRecognitionUsefulNotes(value: string | null | undefined, maxLength = 5000): string | null {
  const text = trimMax(value ?? "", maxLength);
  if (!text || isAIRecognitionProcessNote(text)) return null;
  const withoutAdvice = stripAIRecognitionAdviceClauses(text);
  if (!withoutAdvice || isAIRecognitionProcessNote(withoutAdvice) || isAIRecognitionMarketingNote(withoutAdvice)) return null;
  return trimMax(cleanAIRecognitionServiceDescription(withoutAdvice), maxLength) || null;
}

/** 识别模型推理、低置信提示或“请确认”类备注；这些内容不应随订阅导入长期入库。 */
export function isAIRecognitionProcessNote(value: string): boolean {
  const key = recognitionNoteMatchKey(value);
  return Boolean(key) && AI_RECOGNITION_PROCESS_NOTE_FRAGMENTS.some((fragment) => key.includes(fragment));
}

function recognitionNoteMatchKey(value: string): string {
  return value.normalize("NFKC").trim().toLowerCase().replace(/\s+/g, "");
}

function stripAIRecognitionAdviceClauses(value: string): string {
  const clauses = value.split(/(?<=[。.!?！？；;])|[，,]/u)
    .map((part) => part.trim())
    .filter(Boolean)
    .filter((part) => !AI_RECOGNITION_NOTE_DROP_CLAUSE_FRAGMENTS.some((fragment) => recognitionNoteMatchKey(part).includes(fragment)));
  return clauses.join("，").replace(/，([。.!?！？；;])/g, "$1").trim();
}

function isAIRecognitionMarketingNote(value: string): boolean {
  const key = recognitionNoteMatchKey(value);
  return Boolean(key) && AI_RECOGNITION_NOTE_MARKETING_FRAGMENTS.some((fragment) => key.includes(fragment));
}

function cleanAIRecognitionServiceDescription(value: string): string {
  return value
    .replace(/相关服务/g, "服务")
    .replace(/等服务服务/g, "等服务")
    .replace(/服务服务/g, "服务")
    .replace(/\s+/g, " ")
    .trim();
}

function trimMax(value: string, maxLength: number): string {
  return [...value.trim()].slice(0, maxLength).join("");
}
